Add null masking for parquet decryption

shangxinli opened the pull request:
Summary: Without this change, when the user queries an encrypted parquet file but doesn't have permission for the encryption key, the query will throw an 'access denied' exception. This change is to return a masked value (NULL in the current implementation) for the no-permission-encrypted columns if the added flag is enabled (by default it is disabled). This feature is useful because in some use cases, the user is not allowed to be granted permission for reading some encrypted sensitive column. Another use case is the user doesn't really need the real value of those encrypted columns, but it is to cumbersome forcing user to explicitly write each column they want. Instead, the user just writes 'select * from xxx'. 


Test plan: 

This feature was tested in the Uber environment and then rolled out to production for 2+ years.


```
== RELEASE NOTES ==

General Changes
* Add null masking for the Parquet decryption feature. When this feature is enabled and the user is denied access encrypted column, the columns will be removed in the requested schema sent to Parquet. Then it is filled out with 'NULL'  when the result is returned. 

Hive Changes
* A new session property - read_null_masked_parquet_encrypted_value_enabled is introduced. When this session property is enabled, the Parquet null masking feature is enabled. By default, it is disabled.  
```

yingsu00 reviewed (commented):
In your PR message `A new flag - read_null_masked_parquet_encrypted_value_enabled` did you mean session property? If yes, change `flag` to `session property`. Does it have a config property too? If yes include it in the release message.

shangxinli commented:
> In your PR message `A new flag - read_null_masked_parquet_encrypted_value_enabled` did you mean session property? If yes, change `flag` to `session property`. Does it have a config property too? If yes include it in the release message.

Thanks. I just modified it.

shangxinli commented:
@zhenxiao @beinan Can you have a look?

beinan commented on presto-parquet/src/main/java/com/facebook/presto/parquet/cache/MetadataReader.java:
```diff
@@ -273,7 +284,7 @@ private static ColumnMetaData decryptMetadata(RowGroup rowGroup, ColumnCryptoMet
             return Util.readColumnMetaData(tempInputStream, columnDecryptionSetup.getMetaDataDecryptor(), columnMetaDataAAD);
         }
-        catch (IOException e) {
+        catch (Exception e) {
```
It might be not a good idea to cache all the exceptions, is it possible to cache the exception more explicitly?

shangxinli replied:
That is true. Change it back.

beinan commented on presto-parquet/src/main/java/org/apache/parquet/crypto/ParquetCryptoMetaDataUtils.java:
```diff
@@ -0,0 +1,82 @@
+    {
+        String fieldName = field.getName();
+        currentPath.add(fieldName);
+        ColumnPath path = ColumnPath.get(currentPath.toArray(new String[0]));
```
is it better to use the size of the list to save one array allocation?
```suggestion
        ColumnPath path = ColumnPath.get(currentPath.toArray(new String[currentPath.size()]));
```

zhenxiao replied:
feel same with @beinan

beinan commented on presto-parquet/src/main/java/org/apache/parquet/crypto/ParquetCryptoMetaDataUtils.java:
```diff
@@ -0,0 +1,82 @@
+            }
+        }
+
+        checkArgument(currentPath.size() > 0,
+                "The length of currentPath is empty but trying to remove element in it");
+        checkArgument(currentPath.get(currentPath.size() - 1) != null,
+                "The last element of currentPath is null");
+        checkArgument(currentPath.get(currentPath.size() - 1).equals(fieldName),
+                "The last element of currentPath " + currentPath.get(currentPath.size() - 1) + " not equal to " + fieldName);
+
```
nit: is it better to use checkState?

shangxinli replied:
sure

beinan commented on presto-parquet/src/main/java/org/apache/parquet/crypto/ParquetCryptoMetaDataUtils.java:
```diff
@@ -0,0 +1,82 @@
+        checkArgument(currentPath.get(currentPath.size() - 1).equals(fieldName),
+                "The last element of currentPath " + currentPath.get(currentPath.size() - 1) + " not equal to " + fieldName);
+
+        currentPath.remove(currentPath.size() - 1);
```
I'm a little bit confused, are you trying to remove the element you just added in line 55?

shangxinli replied:
This is a tree recursive navigation.  In the recursed method add/remove are paired up. The 'currentPath' grow/shrink as we traverse to track the path. When the tree is traversed completed and the method is returned completed, the current path become empty.

beinan replied:
Gotcha, that make a lot of sense.  ls it better to use LinkedHashSet instead of List? because looks like you call the `contains` method a lot, it might be an issue if the table is super wide.

beinan reviewed (commented):
Many thanks for the contribution!  it looks like a very useful feature for presto.  just a couple of minor issue and comments.

beinan commented:
There's also some conflict, could you rebase?

shangxinli commented:
> There's also some conflict, could you rebase?
Yeah, whenever the HiveClientConfig is touched, it is very easy to get conflicts. When this PR is about to be merged, I will resolve the conflict once. Otherwise, I will need to keep resolving.

shangxinli commented:
@7c00 @vinothchandar @presto-oss Do you have time to have a look?

vinothchandar commented:
@shangxinli thanks for flagging, taking a look.

shangxinli commented:
Thanks @vinothchandar! Any feedback?

vinothchandar commented:
@shangxinli Was travelling, at it now.

vinothchandar commented:
@7c00 @codope @yihua - could you also try to review this?

codope commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -388,6 +395,11 @@ else if (getParquetType(type, fileSchema, useParquetColumnNames, column, tableNa
             }
+            if (e instanceof HiddenColumnException) {
+                message = format("User does not have access to encryption key for encrypted column = %s. If returning 'null' for encrypted " +
+                        "columns is acceptable to your query, please add 'set session hive.read_null_masked_parquet_encrypted_value_enabled=true' before your query", e.getMessage());
```
Instead of `e.getMessage()`, can we actually get the column name?

shangxinli replied:
Nice feedback. Yeah, we should use the column name.

codope reviewed (commented):
@shangxinli This is a very useful feature! Thanks for the contribution.
Left some comments for your consideration.

shangxinli commented on presto-hudi/src/main/java/com/facebook/presto/hudi/HudiParquetPageSources.java:
```diff
@@ -117,7 +117,7 @@ public static ConnectorPageSource createParquetPageSource(
             dataSource = parquetDataSource;
             Optional<InternalFileDecryptor> fileDecryptor = createDecryptor(configuration, path);
-            ParquetMetadata parquetMetadata = hdfsEnvironment.doAs(user, () -> MetadataReader.readFooter(parquetDataSource, fileSize, fileDecryptor).getParquetMetadata());
+            ParquetMetadata parquetMetadata = hdfsEnvironment.doAs(user, () -> MetadataReader.readFooter(parquetDataSource, fileSize, fileDecryptor, false).getParquetMetadata());
```
Make sense! Changed it.

shangxinli commented on presto-parquet/src/test/java/com/facebook/presto/parquet/reader/TestEncryption.java:
```diff
@@ -220,7 +220,7 @@ private void decryptAndValidate(EncryptionTestFile inputFile)
         Optional<InternalFileDecryptor> fileDecryptor = createFileDecryptor();
         ParquetDataSource dataSource = new MockParquetDataSource(new ParquetDataSourceId(path.toString()), fileSize, inputStream);
-        ParquetMetadata parquetMetadata = MetadataReader.readFooter(dataSource, inputFile.getFileSize(), fileDecryptor).getParquetMetadata();
+        ParquetMetadata parquetMetadata = MetadataReader.readFooter(dataSource, inputFile.getFileSize(), fileDecryptor, false).getParquetMetadata();
```
yeah, just added tests

shangxinli commented:
@codope, sorry for the late response. I just addressed the comments. Please have another look. 

There are a few conflicts now(growing as this PR stay longer), and I fix them at once when all the comments are closed.

codope reviewed (approved):
Looks good. Please rebase.

zhenxiao reviewed (commented):
nice work, @shangxinli looks good to me
only one question, the same with Beinan. could you please address it?

shangxinli commented:
> nice work, @shangxinli looks good to me only one question, the same with Beinan. could you please address it?

Thanks @zhenxiao and @beinan! I didn't know how I missed that comments earlier. I just accepted that change and commit it.

zhenxiao reviewed (approved):
looks good to me
@beinan could you please take a look?

beinan reviewed (approved):
lgtm, thanks!

zhenxiao commented:
hi @shangxinli could you please re-run the test, so that all tests passed?
